src/adapter/src/coord.rs

```diff
@@ -2858,6 +2858,37 @@ pub fn serve(
             .expect("missing EpochMillisseconds timestamp")
             .clone();
 
+        // Choose a time at which to boot. This is the time at which we will run
+        // internal migrations.
+        //
+        // This time is usually the current system time, but with protection
+        // against backwards time jumps, even across restarts.
+        let boot_ts = {
+            // WIP: @joe we have a chicken-and-egg problem here. Ideally we'd
+            // only use monotonic_now when we know we're using the Catalog
+            // oracle, and when using the Postgres oracle we'd only advance the
+            // timestamp by one. But we can only know that once we opened the
+            // Catalog.
+            //
+            // This is bad because we're potentially advancing the timestamp by
+            // 5 seconds, even though we don't have to.
+            let boot_ts = catalog_oracle::monotonic_now(now.clone(), previous_ts);
+            info!(%previous_ts, %boot_ts, "determining boot_ts");
+
+            boot_ts
+        };
+
+        // We need to patch up the EpochMilliseconds timestamp, which will in
+        // turn make sure that we initialize timestamp oracles at the `boot_ts`,
+        // which in turn will make sure that the next `boot_ts` is beyond the
+        // current boot_ts.
+        initial_timestamps
+            .entry(Timeline::EpochMilliseconds)
+            .and_modify(|ts| {
+                *ts = std::cmp::max(*ts, boot_ts);
+            })
+            .or_insert(boot_ts);
+
```

Comment:
I think we need to move the code for persisting the timestamp up here so that we persist the timestamp before using it.

```
for (timeline, initial_timestamp) in initial_timestamps {
                    let persistence =
                        CatalogTimestampPersistence::new(timeline.clone(), Arc::clone(&catalog));
                    handle.block_on(Coordinator::ensure_timeline_state_with_initial_time(
                        &timeline,
                        initial_timestamp,
                        coord_now.clone(),
                        timestamp_oracle_impl,
                        persistence,
                        pg_timestamp_oracle_config.clone(),
                        &mut timestamp_oracles,
                    ));
}
```

Reply:
Then we can remove the call to `txn.set_timestamp` in `open.rs`.